drivers/i2s/i2s_mcux_flexcomm.c

```diff
@@ -0,0 +1,965 @@
+
+			stream->dma_cfg.block_count = NUM_DMA_BLOCKS;
+		} else {
+			stream->dma_cfg.block_count = 1;
+		}
+	} else {
+		blk_cfg->dest_address = (uint32_t)&base->FIFOWR;
+		blk_cfg->source_address = (uint32_t)buffer;
+		blk_cfg->block_size = stream->cfg.block_size;
+		stream->dma_cfg.block_count = 1;
+	}
+
+	stream->dma_cfg.head_block = &stream->dma_block[0];
+	stream->dma_cfg.user_data = (void *)dev;
+}
+
+/* This function is executed in the interrupt context */
+static void i2s_mcux_dma_tx_callback(const struct device *dma_dev, void *arg,
+				uint32_t channel, int status)
+{
+	const struct device *dev = (struct device *)arg;
+	struct i2s_mcux_data *dev_data = dev->data;
+	struct stream *stream = &dev_data->tx;
+	void *buffer;
+	int ret;
+
+	LOG_DBG("tx cb: %d", stream->state);
+	ret = k_msgq_get(&stream->out_queue, &buffer, K_NO_WAIT);
+	if (ret == 0) {
+		/* transmission complete. free the buffer */
+		k_mem_slab_free(stream->cfg.mem_slab, &buffer);
+	} else {
+		LOG_ERR("no buffer in output queue for channel %u", channel);
+	}
+
+	/* Received a STOP trigger, terminate TX immediately */
+	if (stream->last_block) {
+		stream->state = I2S_STATE_READY;
+		i2s_mcux_tx_stream_disable(dev, false);
+		LOG_DBG("TX STOPPED");
+		return;
+	}
+
+	switch (stream->state) {
+	case I2S_STATE_RUNNING:
+	case I2S_STATE_STOPPING:
+		/* get the next buffer from queue */
+		ret = k_msgq_get(&stream->in_queue, &buffer, K_NO_WAIT);
+		if (ret == 0) {
+			/* config the DMA */
+			i2s_mcux_config_dma_blocks(dev, I2S_DIR_TX, (uint32_t *)buffer);
+			dma_config(stream->dev_dma, stream->channel,
+				   &stream->dma_cfg);
+			k_msgq_put(&stream->out_queue, &buffer, K_NO_WAIT);
+			dma_start(stream->dev_dma, stream->channel);
+		}
+
+		if (ret || status) {
+			/*
+			 * DMA encountered an error (status != 0)
+			 * or
+			 * No bufers in input queue
+			 */
+			LOG_DBG("DMA status %08x channel %u k_msgq_get ret %d",
+				status, channel, ret);
+			if (stream->state == I2S_STATE_STOPPING) {
+				stream->state = I2S_STATE_READY;
+			} else {
+				stream->state = I2S_STATE_ERROR;
+			}
+			i2s_mcux_tx_stream_disable(dev, false);
+		}
+		break;
+	case I2S_STATE_ERROR:
+		i2s_mcux_tx_stream_disable(dev, true);
+		break;
+	}
+}
+
+static void i2s_mcux_dma_rx_callback(const struct device *dma_dev, void *arg,
+				uint32_t channel, int status)
+{
+	struct device *dev = (struct device *)arg;
+	struct i2s_mcux_data *dev_data = dev->data;
+	struct stream *stream = &dev_data->rx;
+	void *buffer;
+	int ret;
+
+	LOG_DBG("rx cb: %d", stream->state);
+
+	if (status != 0) {
+		stream->state = I2S_STATE_ERROR;
+		i2s_mcux_rx_stream_disable(dev, false);
+		return;
+	}
+
+	switch (stream->state) {
+	case I2S_STATE_STOPPING:
+	case I2S_STATE_RUNNING:
+		/* retrieve buffer from input queue */
+		ret = k_msgq_get(&stream->in_queue, &buffer, K_NO_WAIT);
+		if (ret != 0) {
+			LOG_ERR("get buffer from in_queue %p failed (%d)",
+				&stream->in_queue, ret);
+			break;
+		}
```

Comment:
If we failed to read data from the queue we need to mark the state of the stream as `I2S_STATE_ERROR`. However, this check could be probably replaced with an ASSERT since it cannot happen that that at this point `stream->in_queue` is empty. That would be a bug in the driver.

Reply:
Thanks, I have changed to an assert.